test/e2e/latency.go

```diff
@@ -0,0 +1,277 @@
+		}.AsSelector())
+	expectNoError(err)
+	for k := range createTimestamps {
+		for _, event := range schedEvents.Items {
+			if event.InvolvedObject.Name == k {
+				scheduleTimestamps[k] = event.FirstTimestamp
+				break
+			}
+		}
+	}
+
+	var (
+		scheduleLatencies        = make([]podLatencyData, 0)
+		startLatencies           = make([]podLatencyData, 0)
+		watchLatencies           = make([]podLatencyData, 0)
+		scheduleToWatchLatencies = make([]podLatencyData, 0)
+		e2eLatencies             = make([]podLatencyData, 0)
+	)
+
+	for name, podNode := range nodes {
+		createTs, ok := createTimestamps[name]
+		Expect(ok).To(Equal(true))
+		scheduleTs, ok := scheduleTimestamps[name]
+		Expect(ok).To(Equal(true))
+		runTs, ok := startTimestamps[name]
+		Expect(ok).To(Equal(true))
+		watchTs, ok := watchTimestamps[name]
+		Expect(ok).To(Equal(true))
+
+		var (
+			scheduleLatency        = podLatencyData{name, podNode, scheduleTs.Time.Sub(createTs.Time)}
+			startLatency           = podLatencyData{name, podNode, runTs.Time.Sub(scheduleTs.Time)}
+			watchLatency           = podLatencyData{name, podNode, watchTs.Time.Sub(runTs.Time)}
+			scheduleToWatchLatency = podLatencyData{name, podNode, watchTs.Time.Sub(scheduleTs.Time)}
+			e2eLatency             = podLatencyData{name, podNode, watchTs.Time.Sub(createTs.Time)}
+		)
+
+		scheduleLatencies = append(scheduleLatencies, scheduleLatency)
+		startLatencies = append(startLatencies, startLatency)
+		watchLatencies = append(watchLatencies, watchLatency)
+		scheduleToWatchLatencies = append(scheduleToWatchLatencies, scheduleToWatchLatency)
+		e2eLatencies = append(e2eLatencies, e2eLatency)
+	}
+
+	sort.Sort(latencySlice(scheduleLatencies))
+	sort.Sort(latencySlice(startLatencies))
+	sort.Sort(latencySlice(watchLatencies))
+	sort.Sort(latencySlice(scheduleToWatchLatencies))
+	sort.Sort(latencySlice(e2eLatencies))
+
+	printLatencies(scheduleLatencies, "worst schedule latencies")
+	printLatencies(startLatencies, "worst run-after-schedule latencies")
+	printLatencies(watchLatencies, "worst watch latencies")
+	printLatencies(scheduleToWatchLatencies, "worst scheduled-to-end total latencies")
+	printLatencies(e2eLatencies, "worst e2e total latencies")
+
+	// Test whether e2e pod startup time is acceptable.
+	// TODO: Switch it to 5 seconds once we are sure our tests are passing.
+	podStartupThreshold := 8 * time.Second
+	e2ePodStartupTime50perc := startLatencies[len(startLatencies)/2].Latency
+	e2ePodStartupTime90perc := startLatencies[len(startLatencies)*9/10].Latency
+	e2ePodStartupTime99perc := startLatencies[len(startLatencies)*99/100].Latency
+	Expect(e2ePodStartupTime50perc).To(BeNumerically("<", podStartupThreshold), "Too high pod startup time 50th percentile")
+	Expect(e2ePodStartupTime90perc).To(BeNumerically("<", podStartupThreshold), "Too high pod startup time 90th percentile")
+	Expect(e2ePodStartupTime99perc).To(BeNumerically("<", podStartupThreshold), "Too high pod startup time 99th percentile")
```

Comment:
FYI: I'm doing some refactoring of this is in: https://github.com/kubernetes/kubernetes/pull/14827
It should be more reusable then...